Flutter_TDD_Architecture_Course_notes.md

```diff
@@ -277,3 +277,310 @@ class GetRandomNumberTrivia extends UseCase<NumberTrivia, NoParams> {
+```
+Then we start the `fromJson` method, these methods always carry a `Map<String, dynamic>` argument.
+That it will be like this:
+
+```dart
+group('fromJson', () {
+    test(
+      'should return a valid model when the JSON number is an integer',
+      () async {
+        // arrange
+        final Map<String, dynamic> jsonMap =
+            json.decode(fixture('trivia.json'));
+        // act
+        final result = NumberTriviaModel.fromJson(jsonMap);
+        // assert
+        expect(result, tNumberTriviaModel);
+```
+The test will now fail and then we should implement the same method in the `number_trivia_model.dart` file.
+
+```dart
+class NumberTriviaModel extends NumberTrivia {
+  ...
+  factory NumberTriviaModel.fromJson(Map<String, dynamic> json) {
+    return NumberTriviaModel(
+      text: json['text'],
+      number: json['number'],
+    );
+  }
+}
+```
+Now we move on to the `second` conversion method.
+
+```dart 
+group('toJson', () {
+  test(
+    'should return a JSON map containing the proper data',
+    () async {
+      // act
+      final result = tNumberTriviaModel.toJson();
+      // assert
+      final expectedJsonMap = {
+        "text": "Test Text",
+        "number": 1,
+      };
+      expect(result, expectedJsonMap);
+    },
+  );
+});
+```
+Implement the method in the file `number_trivia_model.dart`.
+
+```dart
+class NumberTriviaModel extends NumberTrivia {
+  ...
+  Map<String, dynamic> toJson() {
+    return {
+      'text': text,
+      'number': number,
+    };
+  }
+}
+```
+Now all the tests **pass**.
+
+#### 5. Contracts of Data Sources
+> Video 5: https://www.youtube.com/watch?v=m_lkZo6CYcs&t=1s
+
+In the `fifth` tutorial an explanation is given about the importance of `contracts`.
+The contracts will allow the files to communicate directly with the **repository**.
+
+Inside the `repositories` folder create `number_trivia_repository_impl.dart`.
+This is the code for the file.
+
+```dart
+import 'package:dartz/dartz.dart';
+
+import '../../../../core/error/failure.dart';
+import '../../domain/entities/number_trivia.dart';
+import '../../domain/repositories/number_trivia_repository.dart';
+
+class NumberTriviaRepositoryImpl implements NumberTriviaRepository {
+  @override
+  Future<Either<Failure, NumberTrivia>> getConcreteNumberTrivia(int number) {
+    // TODO: implement getConcreteNumberTrivia
+    return null;
+  }
+
+  @override
+  Future<Either<Failure, NumberTrivia>> getRandomNumberTrivia() {
+    // TODO: implement getRandomNumberTrivia
+    return null;
+  }
+}
+```
+
+Inside the `platform` folder create a file called `network_info.dart`with the following code.
+
+```dart
+abstract class NetworkInfo {
+  Future<bool> get isConnected;
+}
+```
+#### Remote Data Source
+
+In this class the value returned will be different from the others and will return only a `NumberTriviaModel` value and the errors will be treated as `Exceptions`.
```

Comment:
This sentence is 27 words. It has two instances of the word "**and**". That is usually a sign that you are combining too many thoughts into the same sentence. It's much better for the reader to split it into two or cut back the one large sentence.
e.g:
```suggestion
In this class the returned value will be a `NumberTriviaModel` value and errors will be treated as `Exceptions`.
```